### PR TITLE
Add capability to filter logs based on latency and ability to control…

### PR DESCRIPTION
**Pull request by arunkumarchacko**
… the information that is getting logged

This is done to limit the amount of data that is getting logged and, if enabled, pushed to cloud logging

### CODE REVIEW

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemConfiguration.java [hunk 1]
```diff
@@ -459,6 +459,21 @@ public class GoogleHadoopFileSystemConfiguration {
...
+   * true.
+   */
+  public static final HadoopConfigurationProperty<Long> GCS_TRACE_LOG_TIME_THRESHOLD_MS =
+      new HadoopConfigurationProperty<>("fs.gs.tracelog.time.filter.threshold.ms", 0L);
+
+  /**
+   * Configuration key to control the properties to log for a given trace. This config is applied
+   * only if fs.gs.tracelog.enable is set to true.
+   */
+  public static final HadoopConfigurationProperty<String> GCS_TRACE_LOG_EXCLUDE_PROPERTIES =
```
Comment (singhravidutt) [inline gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemConfiguration.java:474]: Shouldn't this be List/Set?

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStream.java [hunk 2]
```diff
@@ -259,44 +273,63 @@ public int available() throws IOException {
   }
 
   private void readAPITrace(
```
Comment (singhravidutt) [inline gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStream.java:275]: what about have a default value for log level and provide a way to override it for specific APIs?

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStream.java [hunk 3]
```diff
@@ -171,7 +185,7 @@ public synchronized int read(long position, byte[] buf, int offset, int length)
     Stopwatch stopwatch = Stopwatch.createStarted();
 
     int result = super.read(position, buf, offset, length);
-    readAPITrace(POSITIONAL_READ_METHOD, stopwatch, position, offset, length, result);
+    readAPITrace(POSITIONAL_READ_METHOD, stopwatch, position, offset, length, result, Level.FINE);
```
Comment (singhravidutt) [inline gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStream.java:188]: out of curiosity why one read function is using level as `FINE` and other as `INFO`

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemConfiguration.java (hunk 1 above)
Comment (arunkumarchacko) [reply gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemConfiguration.java]: Done

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemConfiguration.java [hunk 4]
```diff
@@ -459,6 +459,21 @@ public class GoogleHadoopFileSystemConfiguration {
   public static final HadoopConfigurationProperty<Boolean> GCS_TRACE_LOG_ENABLE =
       new HadoopConfigurationProperty<>("fs.gs.tracelog.enable", false);
 
+  /**
+   * Configuration key to control trace logging of operations which exceeded the provided threshold.
+   * Set this to 0 to log everything. This config is applied only if fs.gs.tracelog.enable is set to
+   * true.
+   */
+  public static final HadoopConfigurationProperty<Long> GCS_TRACE_LOG_TIME_THRESHOLD_MS =
```
[the remark this answers is not in the document]
Comment (arunkumarchacko) [reply gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemConfiguration.java:468]: Done

File: gcs/CONFIGURATION.md [hunk 5]
```diff
@@ -574,3 +574,32 @@ permissions (not authorized) to execute these requests.
...
+    performed on the Google Cloud Storage.
+
+*   `fs.gs.tracelog.time.filter.threshold.ms` (default: `0`)
+
+    When this configuration is set, only operations that take longer than the 
+    specified duration will be logged. This flag can be used to reduce the size 
+    of the trace logs by only logging operations that take a significant amount 
+    of time. This config is applied only if `fs.gs.tracelog.enable` is set to true.
+
+*   `fs.gs.tracelog.exclude.properties` (not set by default)
```
[the remark this answers is not in the document]
Comment (arunkumarchacko) [reply gcs/CONFIGURATION.md:593]: Having no "Schema" for the properties is simpler and more flexible. i.e. if a new log property is added, no other code need change. I wanted to keep this feature simple.

[the remark this answers is not in the document]
Comment (singhravidutt) [reply gcs/CONFIGURATION.md:593]: As discussed offline, It makes sense to extract all "keys" and put it into enum and later on make sure that whenever we add log entries via trace logging use that enum.

Enum of all keys will act as the super set of values that can be provided via `fs.gs.tracelog.exclude.properties`.

**Review by singhravidutt — APPROVED** [review]

File: gcs/CONFIGURATION.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (arunkumarchacko) [reply gcs/CONFIGURATION.md:593]: ack

**Pull request merged**